uportal-war/src/main/java/org/apereo/portal/layout/dlm/DistributedLayoutManager.java

```diff
@@ -452,9 +453,80 @@ public IUserLayoutNodeDescription addNode(
 
             return node;
         }
+
+        if (LOG.isTraceEnabled()) {
+            LOG.trace("Attempted to add node in impermissible location "
+                    + "with parent " + parentId + " and next sibling " + nextSiblingId
+                    + ". Made no layout change.");
+        }
+
         return null;
     }
 
+    /*
+     * DLM implementation adds to first allowable position among existing children of the parent
+     * node, if any.
+     */
+    public IUserLayoutNodeDescription addNodeInAnyOrder(
+        final IUserLayoutNodeDescription nodeToInsert, final String parentId) {
+
+        final IUserLayoutNodeDescription parentNode = getNode(parentId);
+
+        final String firstAllowableNextSiblingId =
+            firstAllowableNextSiblingId(parentNode, nodeToInsert);
+
+        return addNode(nodeToInsert, parentId, firstAllowableNextSiblingId);
+    }
+
+    /**
+     * Compute the siblingId representing the earliest allowable placement for adding a child node.
+     * @param parent potential parent of the node to be added
+     * @return siblingId representing earliest allowable nextSiblingId for adding a node,
+     * or blank if no such nextSiblingId
+     * @since 5.0.0
+     */
+    private String firstAllowableNextSiblingId(
```

Comment:
Keeping the verb out of it keeps focus on what the method returns? It's not just procedural (calculate the next sibling ID), it's functional (return the next sibling ID). Caller doesn't need to specify the implementation details -- it might calculate it, it might pull it from a cache, it might ...

A more object-oriented approach to this might push this method into the folder-type Node API such that could ask a folder what's the first allowable next sibling ID for an item to be added to it. I'd have done that except DLM already has this procedural approach going on such that a bunch of helper methods would have to come with it or so.